# comparative_trypanosoma_paper
Custom scripts for recreating some of the analyses from the trypanosoma genomic comparisons work.

[1. Genome Completion Pipeline](#1.-genome-completion-pipeline)  
[2. Genome Annotation Pipeline](#2.-genome-annotation-pipeline)  
[3. Edit Alignments for Phylogeny](#3.-edit-alignments-for-phylogeny)  
[Percent Identity Calculation](#percent-identity-calculation)  

## 1. Genome Completion Pipeline ##
## 2. Genome Annotation Pipeline ##
## 3. Edit Alignments for Phylogeny ##

## [Percent Identity Calculation]

Script name: get_perc_id_general.py

Description: Script for calculating percent identity from aligned nucleotide or amino acid sequences.

Usage: `python get_perc_id_general.py fastafile`

Input  
- fastafile alignment in fasta format

Output  
- percent identity


